### PR TITLE
[twins] avoid timeout and support test case

### PR DESCRIPTION
**Pull request by zekun000**
## Motivation

- Avoid timeout and embed the timeout to the RoundProposer, in order to get rid of the timeout we have the following hacks.
  - Calculate timeout based on the partitions, if the leader partitions don't have enough votes, every one time out, else no one timeout.
  - In order to synchronize the rounds, playground doesn't drop timeout messages, and still sends the sync info from proposal if it's dropped.
- Support the TestCase format which is output by the generator.
- Add a test for the timeout from RoundProposer.
- Implement the execution.

### Have you read the [Contributing Guidelines on pull requests](https://github.com/libra/libra/blob/master/CONTRIBUTING.md#pull-requests)?

(Write your answer here.)

## Test Plan

(Share your test plan here. If you changed code, please provide us with clear instructions for verifying that your changes work.)

## Related PRs

(If this PR adds or changes functionality, please take some time to update the docs at https://github.com/libra/website, and link to your PR here.)

### CODE REVIEW

**Review by davidiw — COMMENTED** [review]
we're changing prod code to support a testing case :O

**Comment by zekun000** [comment]
@davidiw if you mean the RoundProposer, yes, but it's common across the codebase when we have test-only feature that's dispatched by config.

File: config/src/config/consensus_config.rs [hunk 1]
```diff
@@ -55,7 +56,7 @@ pub enum ConsensusProposerType {
     // Pre-specified proposers for each round,
     // or default proposer if round proposer not
     // specified
-    RoundProposer(HashMap<Round, AccountAddress>),
+    RoundProposer(RoundProposerConfig),
```
Comment (aching) [inline config/src/config/consensus_config.rs:61]: We should probably add to the comment, this is only for testing and not for prod like you did for RoundProposer below.

File: consensus/src/liveness/round_proposer_election.rs [hunk 2]
```diff
@@ -4,31 +4,47 @@
...
 
-/// The round proposer maps a round to author
+/// The round proposer maps a round to author (Twins only testing)
 pub struct RoundProposer {
     // A pre-defined map specifying proposers per round
     proposers: HashMap<Round, Author>,
     // Default proposer to use if proposer for a round is unspecified.
     // We hardcode this to the first proposer
     default_proposer: Author,
+    // The rounds that're supposed to timeout (when it's in a different partition than the proposer)
```
Comment (aching) [inline consensus/src/liveness/round_proposer_election.rs:17]:
```suggestion
    // The rounds that are supposed to timeout (when it's in a different partition than the proposer)
```

**Review by aching — COMMENTED** [review]
It's an interesting set of hacks to make these tests more reliable.  How are you measuring this is more reliable than it was previously?

**Comment by zekun000** [comment]
it's more reliable in theory, it's either deterministic pass or fail since there's no timeout.

**Pull request closed without merging**